be/ut/dtm0_log_ut.c

```diff
@@ -0,0 +1,298 @@
+		rc = ut_dl_init(&txd[i], &buf[i], i);
+		M0_UT_ASSERT(rc == 0);
+	}
+
+	/* pa[0].st = EX pa[1].st = IP pa[2].st = IP, pyld = valid */
+	ut_dl_set_pa_state(&txd[0].dtd_pg.dtpg_pa[0], M0_DTPS_EXECUTED);
+	m0_mutex_lock(&log->dl_lock);
+	rc = m0_be_dtm0_log_update(log, NULL, &txd[0], &buf[0]);
+	M0_UT_ASSERT(rc == 0);
+	rec[0] = m0_be_dtm0_log_find(log, &txd[0].dtd_id);
+	M0_UT_ASSERT(rec[0]);
+	rc = ut_dl_verify_log_rec(rec[0], 0);
+	M0_UT_ASSERT(rc);
+
+	ut_dl_set_pa_state(&txd[0].dtd_pg.dtpg_pa[1], M0_DTPS_PERSISTENT);
+	ut_dl_set_pa_state(&txd[0].dtd_pg.dtpg_pa[2], M0_DTPS_PERSISTENT);
+	rc = m0_be_dtm0_log_update(log, NULL, &txd[0], &buf[0]);
+	M0_UT_ASSERT(rc == 0);
+	rec[0] = m0_be_dtm0_log_find(log, &txd[0].dtd_id);
+	M0_UT_ASSERT(rec[0]);
+	rc = ut_dl_verify_log_rec(rec[0], 0);
+	M0_UT_ASSERT(rc);
+
+	rc = m0_be_dtm0_log_prune(log, NULL, &txd[0].dtd_id);
+	M0_UT_ASSERT(rc == -EPROTO);
+
+	ut_dl_set_pa_state(&txd[0].dtd_pg.dtpg_pa[0], M0_DTPS_PERSISTENT);
+	rc = m0_be_dtm0_log_update(log, NULL, &txd[0], &buf[0]);
+	M0_UT_ASSERT(rc == 0);
+
+	rc = m0_be_dtm0_log_prune(log, NULL, &txd[0].dtd_id);
+	M0_UT_ASSERT(rc == 0);
+
+	rec[0] = m0_be_dtm0_log_find(log, &txd[0].dtd_id);
+	M0_UT_ASSERT(rec[0] == NULL);
+
+	rc = m0_be_dtm0_log_prune(log, NULL, &txd[0].dtd_id);
+	M0_UT_ASSERT(rc == -ENOENT);
+
+	ut_dl_set_pa_state(&txd[0].dtd_pg.dtpg_pa[0], M0_DTPS_INPROGRESS);
+	ut_dl_set_pa_state(&txd[0].dtd_pg.dtpg_pa[1], M0_DTPS_INPROGRESS);
+	ut_dl_set_pa_state(&txd[0].dtd_pg.dtpg_pa[2], M0_DTPS_PERSISTENT);
+	rc = m0_be_dtm0_log_update(log, NULL, &txd[0], &empty_buf);
+	M0_UT_ASSERT(rc == 0);
+	rec[0] = m0_be_dtm0_log_find(log, &txd[0].dtd_id);
+	M0_UT_ASSERT(rec[0]);
+	rc = ut_dl_verify_log_rec(rec[0], 0);
+	M0_UT_ASSERT(rc);
+
+	ut_dl_set_pa_state(&txd[0].dtd_pg.dtpg_pa[0], M0_DTPS_PERSISTENT);
+	ut_dl_set_pa_state(&txd[0].dtd_pg.dtpg_pa[1], M0_DTPS_PERSISTENT);
+	rc = m0_be_dtm0_log_update(log, NULL, &txd[0], &buf[0]);
+	M0_UT_ASSERT(rc == 0);
+	rec[0] = m0_be_dtm0_log_find(log, &txd[0].dtd_id);
+	M0_UT_ASSERT(rec[0]);
+	rc = ut_dl_verify_log_rec(rec[0], 0);
+	M0_UT_ASSERT(rc);
+	rc = m0_be_dtm0_log_prune(log, NULL, &txd[0].dtd_id);
+	M0_UT_ASSERT(rc == 0);
+	rec[0] = NULL;
+
+	for (i = 0; i < UT_DTM0_LOG_MAX_LOG_REC; ++i) {
+		ut_dl_set_pa_state(&txd[i].dtd_pg.dtpg_pa[0],
+                                   M0_DTPS_PERSISTENT);
+		ut_dl_set_pa_state(&txd[i].dtd_pg.dtpg_pa[1],
+                                   M0_DTPS_PERSISTENT);
+		ut_dl_set_pa_state(&txd[i].dtd_pg.dtpg_pa[2],
+                                   M0_DTPS_PERSISTENT);
+		rc = m0_be_dtm0_log_update(log, NULL, &txd[i], &buf[i]);
+		M0_UT_ASSERT(rc == 0);
+		rec[i] = m0_be_dtm0_log_find(log, &txd[i].dtd_id);
+		M0_UT_ASSERT(rec[i]);
+		rc = ut_dl_verify_log_rec(rec[i], i);
+		M0_UT_ASSERT(rc);
+		rec[i] = NULL;
+	}
+
+	rec[5] = m0_be_dtm0_log_find(log, &txd[5].dtd_id);
+	M0_UT_ASSERT(rec[5]);
+	rc = ut_dl_verify_log_rec(rec[5], 5);
+	M0_UT_ASSERT(rc);
```

Comment:
rc == 0 please, here and everywhere.

Reply:
Done, it is expecting non zero value so did M0_UT_ASSERT(rc != 0).